### PR TITLE
Cleanup namespace validation

### PR DESCRIPTION
**Pull request by apostasie**
Currently, namespace name is only validated in certain circumstances (eg: if we are going to access a container statedir). This may lead to inconsistencies, where we would allow forbidden namespaces in certain scenarios but properly restrict them in other scenarios.

Also, the current set of rules for validation are inadequate, and fail to cover a range of conditions that *will* fail on the filesystem.

So, suggesting:
- we always validate namespace, once, during argument parsing 
- we use the known good validator we have in place already in the store (ValidatePathComponent) instead of the current validator

This is introducing changes. Specifically, it will restrict previously allowed namespaces:
- namespaces that are longer than 255 characters
- namespaces containing protected OS/filesystem patterns (specifically: on windows, `con[0-9]`, etc)

And on the other hand, relax restrictions on certain patterns that were previously forbidden for no apparent reason (tilde, etc).

Most importantly though, note that containerd will *anyhow* identifier-validate the namespace name whenever it touches it... with the much more restrictive`^[A-Za-z0-9]+(?:[._-](?:[A-Za-z0-9]+))*$` - so, this is merely about cleaning things up on our side.

Let me know what you think.

### CODE REVIEW

File: pkg/nsutil/nsutil.go [hunk 1]
```diff
@@ -1,47 +0,0 @@
...
-		return fmt.Errorf("namespace name cannot be empty")
-	}
-
-	// Slash and '$' for POSIX and backslash and '%' for Windows.
-	pathSeparators := "/\\%$"
-	if strings.ContainsAny(nsName, pathSeparators) {
-		return fmt.Errorf("namespace name cannot contain any special characters (%q): %s", pathSeparators, nsName)
-	}
-
-	specialAliases := []string{".", "..", "~"}
```
Comment (AkihiroSuda) [inline pkg/nsutil/nsutil.go:39]: This check should be retained

File: pkg/nsutil/nsutil_test.go
Comment (djdongjin) [inline pkg/nsutil/nsutil_test.go:1]: We should move the tests to `pkg/store/filestore_test.go` unless they're all covered by existing cases there.

Comment (apostasie) [reply pkg/nsutil/nsutil_test.go:1]: Fair.
(Pending CI) #3725 does test filesystem restrictions thoroughly now.

File: cmd/nerdctl/main.go [hunk 2]
```diff
@@ -239,6 +240,9 @@ Config file ($NERDCTL_TOML): %s
 				return fmt.Errorf("invalid cgroup-manager %q (supported values: \"systemd\", \"cgroupfs\", \"none\")", cgroupManager)
 			}
 		}
+		if err = store.ValidatePathComponent(globalOptions.Namespace); err != nil {
```
Comment (djdongjin) [inline cmd/nerdctl/main.go:243]: nit: 

can we add a comment stating something like namespace should be a single valid path component.
 
https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#path-segment-names

**Review by djdongjin — APPROVED** [review]
LGTM thanks

**Review by djdongjin — APPROVED** [review]

File: cmd/nerdctl/main.go (hunk 2 above)
Comment (apostasie) [reply cmd/nerdctl/main.go:243]: Sure. Let me do that real quick.

**Comment by apostasie** [comment]
Thanks @djdongjin 

Latest push addresses your comment, and has been rebased to account for newly added unit tests.

**Comment by apostasie** [comment]
`TestComposeUp` failure seems unrelated.

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**